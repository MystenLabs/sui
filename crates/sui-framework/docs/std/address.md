---
title: Module `std::address`
---

Provides a way to get address length since it's a
platform-specific parameter.


-  [Function `length`](#std_address_length)


<pre><code></code></pre>



<a name="std_address_length"></a>

## Function `length`

Should be converted to a native function.
Current implementation only works for Sui.


<pre><code><b>public</b> <b>fun</b> <a href="../std/address.md#std_address_length">length</a>(): <a href="../std/u64.md#std_u64">u64</a>
</code></pre>



<details>
<summary>Implementation</summary>


<pre><code><b>public</b> <b>fun</b> <a href="../std/address.md#std_address_length">length</a>(): <a href="../std/u64.md#std_u64">u64</a> {
    32
}
</code></pre>



</details>
