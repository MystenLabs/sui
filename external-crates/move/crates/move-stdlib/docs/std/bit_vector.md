
<a name="std_bit_vector"></a>

# Module `std::bit_vector`



-  [Struct `BitVector`](#std_bit_vector_BitVector)
-  [Constants](#@Constants_0)
-  [Function `new`](#std_bit_vector_new)
-  [Function `set`](#std_bit_vector_set)
-  [Function `unset`](#std_bit_vector_unset)
-  [Function `shift_left`](#std_bit_vector_shift_left)
-  [Function `is_index_set`](#std_bit_vector_is_index_set)
-  [Function `length`](#std_bit_vector_length)
-  [Function `longest_set_sequence_starting_at`](#std_bit_vector_longest_set_sequence_starting_at)


<pre><code></code></pre>



<a name="std_bit_vector_BitVector"></a>

## Struct `BitVector`



<pre><code><b>public</b> <b>struct</b> <a href="../std/bit_vector.md#std_bit_vector_BitVector">BitVector</a> <b>has</b> <b>copy</b>, drop, store
</code></pre>



<details>
<summary>Fields</summary>


<dl>
<dt>
<code><a href="../std/bit_vector.md#std_bit_vector_length">length</a>: <a href="../std/u64.md#std_u64">u64</a></code>
</dt>
<dd>
</dd>
<dt>
<code>bit_field: <a href="../std/vector.md#std_vector">vector</a>&lt;bool&gt;</code>
</dt>
<dd>
</dd>
</dl>


</details>

<a name="@Constants_0"></a>

## Constants


<a name="std_bit_vector_EINDEX"></a>

The provided index is out of bounds


<pre><code><b>const</b> <a href="../std/bit_vector.md#std_bit_vector_EINDEX">EINDEX</a>: <a href="../std/u64.md#std_u64">u64</a> = 131072;
</code></pre>



<a name="std_bit_vector_ELENGTH"></a>

An invalid length of bitvector was given


<pre><code><b>const</b> <a href="../std/bit_vector.md#std_bit_vector_ELENGTH">ELENGTH</a>: <a href="../std/u64.md#std_u64">u64</a> = 131073;
</code></pre>



<a name="std_bit_vector_MAX_SIZE"></a>

The maximum allowed bitvector size


<pre><code><b>const</b> <a href="../std/bit_vector.md#std_bit_vector_MAX_SIZE">MAX_SIZE</a>: <a href="../std/u64.md#std_u64">u64</a> = 1024;
</code></pre>



<a name="std_bit_vector_WORD_SIZE"></a>



<pre><code><b>const</b> <a href="../std/bit_vector.md#std_bit_vector_WORD_SIZE">WORD_SIZE</a>: <a href="../std/u64.md#std_u64">u64</a> = 1;
</code></pre>



<a name="std_bit_vector_new"></a>

## Function `new`



<pre><code><b>public</b> <b>fun</b> <a href="../std/bit_vector.md#std_bit_vector_new">new</a>(<a href="../std/bit_vector.md#std_bit_vector_length">length</a>: <a href="../std/u64.md#std_u64">u64</a>): <a href="../std/bit_vector.md#std_bit_vector_BitVector">std::bit_vector::BitVector</a>
</code></pre>



<details>
<summary>Implementation</summary>


<pre><code><b>public</b> <b>fun</b> <a href="../std/bit_vector.md#std_bit_vector_new">new</a>(<a href="../std/bit_vector.md#std_bit_vector_length">length</a>: <a href="../std/u64.md#std_u64">u64</a>): <a href="../std/bit_vector.md#std_bit_vector_BitVector">BitVector</a> {
    <b>assert</b>!(<a href="../std/bit_vector.md#std_bit_vector_length">length</a> &gt; 0, <a href="../std/bit_vector.md#std_bit_vector_ELENGTH">ELENGTH</a>);
    <b>assert</b>!(<a href="../std/bit_vector.md#std_bit_vector_length">length</a> &lt; <a href="../std/bit_vector.md#std_bit_vector_MAX_SIZE">MAX_SIZE</a>, <a href="../std/bit_vector.md#std_bit_vector_ELENGTH">ELENGTH</a>);
    <b>let</b> <b>mut</b> counter = 0;
    <b>let</b> <b>mut</b> bit_field = <a href="../std/vector.md#std_vector_empty">vector::empty</a>();
    <b>while</b> (counter &lt; <a href="../std/bit_vector.md#std_bit_vector_length">length</a>) {
        bit_field.push_back(<b>false</b>);
        counter = counter + 1;
    };
    <a href="../std/bit_vector.md#std_bit_vector_BitVector">BitVector</a> {
        <a href="../std/bit_vector.md#std_bit_vector_length">length</a>,
        bit_field,
    }
}
</code></pre>



</details>

<a name="std_bit_vector_set"></a>

## Function `set`

Set the bit at <code>bit_index</code> in the <code>bitvector</code> regardless of its previous state.


<pre><code><b>public</b> <b>fun</b> <a href="../std/bit_vector.md#std_bit_vector_set">set</a>(bitvector: &<b>mut</b> <a href="../std/bit_vector.md#std_bit_vector_BitVector">std::bit_vector::BitVector</a>, bit_index: <a href="../std/u64.md#std_u64">u64</a>)
</code></pre>



<details>
<summary>Implementation</summary>


<pre><code><b>public</b> <b>fun</b> <a href="../std/bit_vector.md#std_bit_vector_set">set</a>(bitvector: &<b>mut</b> <a href="../std/bit_vector.md#std_bit_vector_BitVector">BitVector</a>, bit_index: <a href="../std/u64.md#std_u64">u64</a>) {
    <b>assert</b>!(bit_index &lt; bitvector.bit_field.<a href="../std/bit_vector.md#std_bit_vector_length">length</a>(), <a href="../std/bit_vector.md#std_bit_vector_EINDEX">EINDEX</a>);
    <b>let</b> x = &<b>mut</b> bitvector.bit_field[bit_index];
    *x = <b>true</b>;
}
</code></pre>



</details>

<a name="std_bit_vector_unset"></a>

## Function `unset`

Unset the bit at <code>bit_index</code> in the <code>bitvector</code> regardless of its previous state.


<pre><code><b>public</b> <b>fun</b> <a href="../std/bit_vector.md#std_bit_vector_unset">unset</a>(bitvector: &<b>mut</b> <a href="../std/bit_vector.md#std_bit_vector_BitVector">std::bit_vector::BitVector</a>, bit_index: <a href="../std/u64.md#std_u64">u64</a>)
</code></pre>



<details>
<summary>Implementation</summary>


<pre><code><b>public</b> <b>fun</b> <a href="../std/bit_vector.md#std_bit_vector_unset">unset</a>(bitvector: &<b>mut</b> <a href="../std/bit_vector.md#std_bit_vector_BitVector">BitVector</a>, bit_index: <a href="../std/u64.md#std_u64">u64</a>) {
    <b>assert</b>!(bit_index &lt; bitvector.bit_field.<a href="../std/bit_vector.md#std_bit_vector_length">length</a>(), <a href="../std/bit_vector.md#std_bit_vector_EINDEX">EINDEX</a>);
    <b>let</b> x = &<b>mut</b> bitvector.bit_field[bit_index];
    *x = <b>false</b>;
}
</code></pre>



</details>

<a name="std_bit_vector_shift_left"></a>

## Function `shift_left`

Shift the <code>bitvector</code> left by <code>amount</code>. If <code>amount</code> is greater than the
bitvector's length the bitvector will be zeroed out.


<pre><code><b>public</b> <b>fun</b> <a href="../std/bit_vector.md#std_bit_vector_shift_left">shift_left</a>(bitvector: &<b>mut</b> <a href="../std/bit_vector.md#std_bit_vector_BitVector">std::bit_vector::BitVector</a>, amount: <a href="../std/u64.md#std_u64">u64</a>)
</code></pre>



<details>
<summary>Implementation</summary>


<pre><code><b>public</b> <b>fun</b> <a href="../std/bit_vector.md#std_bit_vector_shift_left">shift_left</a>(bitvector: &<b>mut</b> <a href="../std/bit_vector.md#std_bit_vector_BitVector">BitVector</a>, amount: <a href="../std/u64.md#std_u64">u64</a>) {
    <b>if</b> (amount &gt;= bitvector.<a href="../std/bit_vector.md#std_bit_vector_length">length</a>) {
       <b>let</b> len = bitvector.bit_field.<a href="../std/bit_vector.md#std_bit_vector_length">length</a>();
       <b>let</b> <b>mut</b> i = 0;
       <b>while</b> (i &lt; len) {
           <b>let</b> elem = &<b>mut</b> bitvector.bit_field[i];
           *elem = <b>false</b>;
           i = i + 1;
       };
    } <b>else</b> {
        <b>let</b> <b>mut</b> i = amount;
        <b>while</b> (i &lt; bitvector.<a href="../std/bit_vector.md#std_bit_vector_length">length</a>) {
            <b>if</b> (bitvector.<a href="../std/bit_vector.md#std_bit_vector_is_index_set">is_index_set</a>(i)) bitvector.<a href="../std/bit_vector.md#std_bit_vector_set">set</a>(i - amount)
            <b>else</b> bitvector.<a href="../std/bit_vector.md#std_bit_vector_unset">unset</a>(i - amount);
            i = i + 1;
        };
        i = bitvector.<a href="../std/bit_vector.md#std_bit_vector_length">length</a> - amount;
        <b>while</b> (i &lt; bitvector.<a href="../std/bit_vector.md#std_bit_vector_length">length</a>) {
            <a href="../std/bit_vector.md#std_bit_vector_unset">unset</a>(bitvector, i);
            i = i + 1;
        };
    }
}
</code></pre>



</details>

<a name="std_bit_vector_is_index_set"></a>

## Function `is_index_set`

Return the value of the bit at <code>bit_index</code> in the <code>bitvector</code>. <code><b>true</b></code>
represents "1" and <code><b>false</b></code> represents a 0


<pre><code><b>public</b> <b>fun</b> <a href="../std/bit_vector.md#std_bit_vector_is_index_set">is_index_set</a>(bitvector: &<a href="../std/bit_vector.md#std_bit_vector_BitVector">std::bit_vector::BitVector</a>, bit_index: <a href="../std/u64.md#std_u64">u64</a>): bool
</code></pre>



<details>
<summary>Implementation</summary>


<pre><code><b>public</b> <b>fun</b> <a href="../std/bit_vector.md#std_bit_vector_is_index_set">is_index_set</a>(bitvector: &<a href="../std/bit_vector.md#std_bit_vector_BitVector">BitVector</a>, bit_index: <a href="../std/u64.md#std_u64">u64</a>): bool {
    <b>assert</b>!(bit_index &lt; bitvector.bit_field.<a href="../std/bit_vector.md#std_bit_vector_length">length</a>(), <a href="../std/bit_vector.md#std_bit_vector_EINDEX">EINDEX</a>);
    bitvector.bit_field[bit_index]
}
</code></pre>



</details>

<a name="std_bit_vector_length"></a>

## Function `length`

Return the length (number of usable bits) of this bitvector


<pre><code><b>public</b> <b>fun</b> <a href="../std/bit_vector.md#std_bit_vector_length">length</a>(bitvector: &<a href="../std/bit_vector.md#std_bit_vector_BitVector">std::bit_vector::BitVector</a>): <a href="../std/u64.md#std_u64">u64</a>
</code></pre>



<details>
<summary>Implementation</summary>


<pre><code><b>public</b> <b>fun</b> <a href="../std/bit_vector.md#std_bit_vector_length">length</a>(bitvector: &<a href="../std/bit_vector.md#std_bit_vector_BitVector">BitVector</a>): <a href="../std/u64.md#std_u64">u64</a> {
    bitvector.bit_field.<a href="../std/bit_vector.md#std_bit_vector_length">length</a>()
}
</code></pre>



</details>

<a name="std_bit_vector_longest_set_sequence_starting_at"></a>

## Function `longest_set_sequence_starting_at`

Returns the length of the longest sequence of set bits starting at (and
including) <code>start_index</code> in the <code>bitvector</code>. If there is no such
sequence, then <code>0</code> is returned.


<pre><code><b>public</b> <b>fun</b> <a href="../std/bit_vector.md#std_bit_vector_longest_set_sequence_starting_at">longest_set_sequence_starting_at</a>(bitvector: &<a href="../std/bit_vector.md#std_bit_vector_BitVector">std::bit_vector::BitVector</a>, start_index: <a href="../std/u64.md#std_u64">u64</a>): <a href="../std/u64.md#std_u64">u64</a>
</code></pre>



<details>
<summary>Implementation</summary>


<pre><code><b>public</b> <b>fun</b> <a href="../std/bit_vector.md#std_bit_vector_longest_set_sequence_starting_at">longest_set_sequence_starting_at</a>(bitvector: &<a href="../std/bit_vector.md#std_bit_vector_BitVector">BitVector</a>, start_index: <a href="../std/u64.md#std_u64">u64</a>): <a href="../std/u64.md#std_u64">u64</a> {
    <b>assert</b>!(start_index &lt; bitvector.<a href="../std/bit_vector.md#std_bit_vector_length">length</a>, <a href="../std/bit_vector.md#std_bit_vector_EINDEX">EINDEX</a>);
    <b>let</b> <b>mut</b> index = start_index;
    // Find the greatest index in the <a href="../std/vector.md#std_vector">vector</a> such that all indices less than it are <a href="../std/bit_vector.md#std_bit_vector_set">set</a>.
    <b>while</b> (index &lt; bitvector.<a href="../std/bit_vector.md#std_bit_vector_length">length</a>) {
        <b>if</b> (!bitvector.<a href="../std/bit_vector.md#std_bit_vector_is_index_set">is_index_set</a>(index)) <b>break</b>;
        index = index + 1;
    };
    index - start_index
}
</code></pre>



</details>


[//]: # ("File containing references which can be used from documentation")
