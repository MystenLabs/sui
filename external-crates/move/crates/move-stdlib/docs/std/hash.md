
<a name="std_hash"></a>

# Module `std::hash`

Module which defines SHA hashes for byte vectors.

The functions in this module are natively declared both in the Move runtime
as in the Move prover's prelude.


-  [Function `sha2_256`](#std_hash_sha2_256)
-  [Function `sha3_256`](#std_hash_sha3_256)


<pre><code></code></pre>



<a name="std_hash_sha2_256"></a>

## Function `sha2_256`



<pre><code><b>public</b> <b>fun</b> <a href="../std/hash.md#std_hash_sha2_256">sha2_256</a>(data: <a href="../std/vector.md#std_vector">vector</a>&lt;<a href="../std/u8.md#std_u8">u8</a>&gt;): <a href="../std/vector.md#std_vector">vector</a>&lt;<a href="../std/u8.md#std_u8">u8</a>&gt;
</code></pre>



<details>
<summary>Implementation</summary>


<pre><code><b>native</b> <b>public</b> <b>fun</b> <a href="../std/hash.md#std_hash_sha2_256">sha2_256</a>(data: <a href="../std/vector.md#std_vector">vector</a>&lt;<a href="../std/u8.md#std_u8">u8</a>&gt;): <a href="../std/vector.md#std_vector">vector</a>&lt;<a href="../std/u8.md#std_u8">u8</a>&gt;;
</code></pre>



</details>

<a name="std_hash_sha3_256"></a>

## Function `sha3_256`



<pre><code><b>public</b> <b>fun</b> <a href="../std/hash.md#std_hash_sha3_256">sha3_256</a>(data: <a href="../std/vector.md#std_vector">vector</a>&lt;<a href="../std/u8.md#std_u8">u8</a>&gt;): <a href="../std/vector.md#std_vector">vector</a>&lt;<a href="../std/u8.md#std_u8">u8</a>&gt;
</code></pre>



<details>
<summary>Implementation</summary>


<pre><code><b>native</b> <b>public</b> <b>fun</b> <a href="../std/hash.md#std_hash_sha3_256">sha3_256</a>(data: <a href="../std/vector.md#std_vector">vector</a>&lt;<a href="../std/u8.md#std_u8">u8</a>&gt;): <a href="../std/vector.md#std_vector">vector</a>&lt;<a href="../std/u8.md#std_u8">u8</a>&gt;;
</code></pre>



</details>


[//]: # ("File containing references which can be used from documentation")
