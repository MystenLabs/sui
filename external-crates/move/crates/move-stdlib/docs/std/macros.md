
<a name="std_macros"></a>

# Module `std::macros`

This module holds shared implementation of macros used in <code>std</code>




<pre><code></code></pre>


[//]: # ("File containing references which can be used from documentation")
