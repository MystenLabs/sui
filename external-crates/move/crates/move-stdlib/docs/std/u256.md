
<a name="std_u256"></a>

# Module `std::u256`



-  [Function `max`](#std_u256_max)
-  [Function `min`](#std_u256_min)
-  [Function `diff`](#std_u256_diff)
-  [Function `divide_and_round_up`](#std_u256_divide_and_round_up)
-  [Function `pow`](#std_u256_pow)


<pre><code></code></pre>



<a name="std_u256_max"></a>

## Function `max`

Return the larger of <code>x</code> and <code>y</code>


<pre><code><b>public</b> <b>fun</b> <a href="../std/u256.md#std_u256_max">max</a>(x: <a href="../std/u256.md#std_u256">u256</a>, y: <a href="../std/u256.md#std_u256">u256</a>): <a href="../std/u256.md#std_u256">u256</a>
</code></pre>



<details>
<summary>Implementation</summary>


<pre><code><b>public</b> <b>fun</b> <a href="../std/u256.md#std_u256_max">max</a>(x: <a href="../std/u256.md#std_u256">u256</a>, y: <a href="../std/u256.md#std_u256">u256</a>): <a href="../std/u256.md#std_u256">u256</a> {
    <a href="../std/macros.md#std_macros_num_max">std::macros::num_max</a>!(x, y)
}
</code></pre>



</details>

<a name="std_u256_min"></a>

## Function `min`

Return the smaller of <code>x</code> and <code>y</code>


<pre><code><b>public</b> <b>fun</b> <a href="../std/u256.md#std_u256_min">min</a>(x: <a href="../std/u256.md#std_u256">u256</a>, y: <a href="../std/u256.md#std_u256">u256</a>): <a href="../std/u256.md#std_u256">u256</a>
</code></pre>



<details>
<summary>Implementation</summary>


<pre><code><b>public</b> <b>fun</b> <a href="../std/u256.md#std_u256_min">min</a>(x: <a href="../std/u256.md#std_u256">u256</a>, y: <a href="../std/u256.md#std_u256">u256</a>): <a href="../std/u256.md#std_u256">u256</a> {
    <a href="../std/macros.md#std_macros_num_min">std::macros::num_min</a>!(x, y)
}
</code></pre>



</details>

<a name="std_u256_diff"></a>

## Function `diff`

Return the absolute value of x - y


<pre><code><b>public</b> <b>fun</b> <a href="../std/u256.md#std_u256_diff">diff</a>(x: <a href="../std/u256.md#std_u256">u256</a>, y: <a href="../std/u256.md#std_u256">u256</a>): <a href="../std/u256.md#std_u256">u256</a>
</code></pre>



<details>
<summary>Implementation</summary>


<pre><code><b>public</b> <b>fun</b> <a href="../std/u256.md#std_u256_diff">diff</a>(x: <a href="../std/u256.md#std_u256">u256</a>, y: <a href="../std/u256.md#std_u256">u256</a>): <a href="../std/u256.md#std_u256">u256</a> {
    <a href="../std/macros.md#std_macros_num_diff">std::macros::num_diff</a>!(x, y)
}
</code></pre>



</details>

<a name="std_u256_divide_and_round_up"></a>

## Function `divide_and_round_up`

Calculate x / y, but round up the result.


<pre><code><b>public</b> <b>fun</b> <a href="../std/u256.md#std_u256_divide_and_round_up">divide_and_round_up</a>(x: <a href="../std/u256.md#std_u256">u256</a>, y: <a href="../std/u256.md#std_u256">u256</a>): <a href="../std/u256.md#std_u256">u256</a>
</code></pre>



<details>
<summary>Implementation</summary>


<pre><code><b>public</b> <b>fun</b> <a href="../std/u256.md#std_u256_divide_and_round_up">divide_and_round_up</a>(x: <a href="../std/u256.md#std_u256">u256</a>, y: <a href="../std/u256.md#std_u256">u256</a>): <a href="../std/u256.md#std_u256">u256</a> {
    <a href="../std/macros.md#std_macros_num_divide_and_round_up">std::macros::num_divide_and_round_up</a>!(x, y)
}
</code></pre>



</details>

<a name="std_u256_pow"></a>

## Function `pow`

Return the value of a base raised to a power


<pre><code><b>public</b> <b>fun</b> <a href="../std/u256.md#std_u256_pow">pow</a>(base: <a href="../std/u256.md#std_u256">u256</a>, exponent: <a href="../std/u8.md#std_u8">u8</a>): <a href="../std/u256.md#std_u256">u256</a>
</code></pre>



<details>
<summary>Implementation</summary>


<pre><code><b>public</b> <b>fun</b> <a href="../std/u256.md#std_u256_pow">pow</a>(base: <a href="../std/u256.md#std_u256">u256</a>, exponent: <a href="../std/u8.md#std_u8">u8</a>): <a href="../std/u256.md#std_u256">u256</a> {
    <a href="../std/macros.md#std_macros_num_pow">std::macros::num_pow</a>!(base, exponent)
}
</code></pre>



</details>


[//]: # ("File containing references which can be used from documentation")
