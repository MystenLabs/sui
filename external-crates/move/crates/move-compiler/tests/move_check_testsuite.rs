// Copyright (c) The Diem Core Contributors
// Copyright (c) The Move Contributors
// SPDX-License-Identifier: Apache-2.0

use std::{collections::BTreeMap, fs, path::Path};

use move_command_line_common::{
    env::read_bool_env_var,
    testing::{add_update_baseline_fix, format_diff, read_env_update_baseline, EXP_EXT, OUT_EXT},
};
use move_compiler::{
    command_line::compiler::move_check_for_errors,
    diagnostics::*,
    editions::{Edition, Flavor},
    linters::{self, LintLevel},
    shared::{Flags, NumericalAddress, PackageConfig, PackagePaths},
    sui_mode, Compiler, PASS_PARSER,
};

/// Shared flag to keep any temporary results of the test
const KEEP_TMP: &str = "KEEP";

const TEST_EXT: &str = "unit_test";
const UNUSED_EXT: &str = "unused";
const MIGRATION_EXT: &str = "migration";
const IDE_EXT: &str = "ide";

const LINTER_DIR: &str = "linter";
const SUI_MODE_DIR: &str = "sui_mode";
const MOVE_2024_DIR: &str = "move_2024";
const DEV_DIR: &str = "development";

fn default_testing_addresses(flavor: Flavor) -> BTreeMap<String, NumericalAddress> {
    let mut mapping = vec![
        ("std", "0x1"),
        ("sui", "0x2"),
        ("M", "0x40"),
        ("A", "0x41"),
        ("B", "0x42"),
        ("K", "0x19"),
        ("a", "0x44"),
        ("b", "0x45"),
        ("k", "0x19"),
    ];
    if flavor == Flavor::Sui {
        mapping.extend([("sui", "0x2"), ("sui_system", "0x3")]);
    }
    mapping
        .into_iter()
        .map(|(name, addr)| (name.to_string(), NumericalAddress::parse_str(addr).unwrap()))
        .collect()
}

fn move_check_testsuite(path: &Path) -> datatest_stable::Result<()> {
    let path_contains = |s| path.components().any(|c| c.as_os_str() == s);
    let lint = path_contains(LINTER_DIR);
    let flavor = if path_contains(SUI_MODE_DIR) {
        Flavor::Sui
    } else {
        Flavor::default()
    };
    let edition = if path_contains(MOVE_2024_DIR) {
        Edition::E2024_ALPHA
    } else if path_contains(DEV_DIR) {
        Edition::DEVELOPMENT
    } else {
        Edition::default()
    };
    let config = PackageConfig {
        flavor,
        edition,
        ..PackageConfig::default()
    };
    testsuite(path, config, lint)
}

fn testsuite(path: &Path, mut config: PackageConfig, lint: bool) -> datatest_stable::Result<()> {
    // A test is marked that it should also be compiled in test mode by having a `path.unit_test`
    // file.
    if path.with_extension(TEST_EXT).exists() {
        let test_exp_path = format!(
            "{}.{TEST_EXT}.{EXP_EXT}",
            path.with_extension("").to_string_lossy(),
        );
        let test_out_path = format!(
            "{}.{TEST_EXT}.{OUT_EXT}",
            path.with_extension("").to_string_lossy(),
        );
        let mut config = config.clone();
        config
            .warning_filter
            .union(&WarningFilters::unused_warnings_filter_for_test());
        run_test(
            path,
            Path::new(&test_exp_path),
            Path::new(&test_out_path),
            Flags::testing(),
            config,
            lint,
        )?;
    }

    // A test is marked that it should also be compiled in migration mode by having a
    // `path.migration` file.
    if path.with_extension(MIGRATION_EXT).exists() {
        let migration_exp_path = format!(
            "{}.{MIGRATION_EXT}.{EXP_EXT}",
            path.with_extension("").to_string_lossy(),
        );
        let migration_out_path = format!(
            "{}.{MIGRATION_EXT}.{OUT_EXT}",
            path.with_extension("").to_string_lossy(),
        );
        let mut config = config.clone();
        config
            .warning_filter
            .union(&WarningFilters::unused_warnings_filter_for_test());
        run_test_inner(
            path,
            Path::new(&migration_exp_path),
            Path::new(&migration_out_path),
            Flags::testing(),
            config,
            lint,
            true,
        )?;
    }

    // A cross-module unused case that should run without unused warnings suppression
    if path.with_extension(UNUSED_EXT).exists() {
        let unused_exp_path = format!(
            "{}.{UNUSED_EXT}.{EXP_EXT}",
            path.with_extension("").to_string_lossy(),
        );
        let unused_out_path = format!(
            "{}.{UNUSED_EXT}.{OUT_EXT}",
            path.with_extension("").to_string_lossy(),
        );
        run_test(
            path,
            Path::new(&unused_exp_path),
            Path::new(&unused_out_path),
            Flags::testing(),
            config.clone(),
            lint,
        )?;
    }

    // A cross-module unused case that should run without unused warnings suppression
    if path.with_extension(IDE_EXT).exists() {
        let ide_exp_path = format!(
            "{}.{IDE_EXT}.{EXP_EXT}",
            path.with_extension("").to_string_lossy(),
        );
        let ide_out_path = format!(
            "{}.{IDE_EXT}.{OUT_EXT}",
            path.with_extension("").to_string_lossy(),
        );
        run_test(
            path,
            Path::new(&ide_exp_path),
            Path::new(&ide_out_path),
            Flags::testing().set_ide_test_mode(true).set_ide_mode(true),
            config.clone(),
            lint,
        )?;
    }

    let exp_path = path.with_extension(EXP_EXT);
    let out_path = path.with_extension(OUT_EXT);

    config
        .warning_filter
        .union(&WarningFilters::unused_warnings_filter_for_test());
    run_test(path, &exp_path, &out_path, Flags::empty(), config, lint)?;
    Ok(())
}

// Runs all tests under the test/testsuite directory.
pub fn run_test(
    path: &Path,
    exp_path: &Path,
    out_path: &Path,
    flags: Flags,
    default_config: PackageConfig,
    lint: bool,
) -> anyhow::Result<()> {
    run_test_inner(path, exp_path, out_path, flags, default_config, lint, false)
}

// Runs all tests under the test/testsuite directory.
pub fn run_test_inner(
    path: &Path,
    exp_path: &Path,
    out_path: &Path,
    flags: Flags,
    package_config: PackageConfig,
    lint: bool,
    migration_mode: bool,
) -> anyhow::Result<()> {
    let flavor = package_config.flavor;
    let targets: Vec<String> = vec![path.to_str().unwrap().to_owned()];
    let named_address_map = default_testing_addresses(flavor);
    let deps = vec![PackagePaths {
        name: Some(("stdlib".into(), PackageConfig::default())),
        paths: move_stdlib::move_stdlib_files(),
        named_address_map: named_address_map.clone(),
    }];
    let name = if migration_mode {
        let mut config = package_config.clone();
        config.edition = Edition::E2024_MIGRATION;
        Some(("test".into(), config))
    } else {
        None
    };
    let targets = vec![PackagePaths {
        name,
        paths: targets,
        named_address_map,
    }];

    let flags = flags.set_sources_shadow_deps(true);

    let mut compiler = Compiler::from_package_paths(None, targets, deps)
        .unwrap()
        .set_flags(flags)
        .set_default_config(package_config);

    if flavor == Flavor::Sui {
        let (prefix, filters) = sui_mode::linters::known_filters();
        compiler = compiler.add_custom_known_filters(prefix, filters);
        if lint {
            compiler = compiler.add_visitors(sui_mode::linters::linter_visitors(LintLevel::All))
        }
    }
    let (prefix, filters) = linters::known_filters();
    compiler = compiler.add_custom_known_filters(prefix, filters);
    if lint {
        compiler = compiler.add_visitors(linters::linter_visitors(LintLevel::All))
    }

    let (files, comments_and_compiler_res) = compiler.run::<PASS_PARSER>()?;
    let diags = move_check_for_errors(comments_and_compiler_res);

    let has_diags = !diags.is_empty();
    let diag_buffer = if has_diags {
        if migration_mode {
            report_migration_to_buffer(&files, diags)
        } else {
            report_diagnostics_to_buffer(&files, diags, /* ansi_color */ false)
        }
    } else {
        vec![]
    };

    let save_diags = read_bool_env_var(KEEP_TMP);
    let update_baseline = read_env_update_baseline();

    let rendered_diags = std::str::from_utf8(&diag_buffer)?;
    if save_diags {
        fs::write(out_path, &diag_buffer)?;
    }

    if update_baseline {
        if has_diags {
            fs::write(exp_path, rendered_diags)?;
        } else if exp_path.is_file() {
            fs::remove_file(exp_path)?;
        }
        return Ok(());
    }

    let exp_exists = exp_path.is_file();
    match (has_diags, exp_exists) {
        (false, false) => Ok(()),
        (true, false) => {
            let msg = format!(
                "Expected success. Unexpected diagnostics:\n{}",
                rendered_diags
            );
            anyhow::bail!(add_update_baseline_fix(msg))
        }
        (false, true) => {
            let msg = format!(
                "Unexpected success. Expected diagnostics:\n{}",
                fs::read_to_string(exp_path)?
            );
            anyhow::bail!(add_update_baseline_fix(msg))
        }
        (true, true) => {
            let expected_diags = fs::read_to_string(exp_path)?;
            if rendered_diags != expected_diags {
                let msg = format!(
                    "Expected diagnostics differ from actual diagnostics:\n{}",
                    format_diff(expected_diags, rendered_diags),
                );
                anyhow::bail!(add_update_baseline_fix(msg))
            } else {
                Ok(())
            }
        }
    }
}

datatest_stable::harness!(move_check_testsuite, "tests/", r".*\.move$");
