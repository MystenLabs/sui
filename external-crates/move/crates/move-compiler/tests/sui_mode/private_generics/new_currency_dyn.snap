---
source: crates/move-compiler/tests/move_check_testsuite.rs
info:
  flavor: sui
  edition: legacy
  lint: false
---
error[Sui E02010]: invalid dynamic coin creation call
   ┌─ tests/sui_mode/private_generics/new_currency_dyn.move:13:5
   │
13 │     coin_registry::new_currency_dyn<External>();
   │     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   │     │                               │
   │     │                               The type 'a::other::External' is not declared in the current module
   │     Invalid dynamic coin creation. The function 'sui::coin_registry::new_currency_dyn' must be called with a type defined in the current module
